onmt/translate/BeamSearcher.lua

```diff
@@ -0,0 +1,298 @@
+    for j = 3, #h:size() do
+        sizes[j - 1] = h:size(j)
+    end
+    return h:view(table.unpack(sizes))
+  end)
+end
+
+--[[ Construct a BeamSearcher object. BeamSearcher takes charge of the beam search related logic such as ranking the predictions, tracking the parents in each beam, and transforming the outputs from last time step according to the parent indexes in the current time step.
+
+Parameters:
+  * `stepFunction` - this function specifies how to go one step forward. It will take a table `stepInputs` as input, produce a table `stepOutputs` as output. All tensors inside tables must have the same first dimension `batchSize`. `stepOutputs[1]` should be of shape (`batchSize`, `numTokens`), which denotes the scores in this step.
+  * `feedFunction` - this function specifies how to prepare the input `stepInputs` to stepFunction given `stepOutputs` and predictions from last time step. All tensors inside tables must have the same first dimension `batchSize`. Input: `topIndexes`, a tensor of shape (`batchSize`), which are the predictions from last time step; `stepOutputs`, the corresponding output of `stepFunction` from last time step. At the first time step, both `stepOutputs` and `topIndexes` will be passed as nil, and the returned values of this function acts as an initializer.
+  * `maxSeqLength` - maximum output sequence length.
+  * `endSymbol` - end of sequence symbol. [onmt.Constants.EOS]
+  * `allowEmptyHyp` - whether or not allow empty sequence. If set to false, then only non-empty sequences will be considered. Otherwise, it may output empty sequences, and there may be nils in the predictions. [false]
+
+Example:
+
+    See function `translateBatch` in `onmt/translate/Translator.lua`. Note that by specifying `stepFunction` and `feedFunction`, we can use it to perform general beam search. We can easily add filters to prune some sequences during beam search, like sequences with too many UNKs.
+
+]]
+function BeamSearcher:__init(stepFunction, feedFunction, maxSeqLength, endSymbol, allowEmptyHyp)
+  self.stepFunction = stepFunction
+  self.feedFunction = feedFunction
+  self.maxSeqLength = maxSeqLength
+  self.endSymbol = endSymbol or onmt.Constants.EOS
+  self.allowEmptyHyp = allowEmptyHyp or false -- by default, we only consider sequences of length >= 1
+end
+
+--[[ Perform beam search.
+
+Parameters:
+  * `beamSize` - this function specifies how to go one step forward. It will take a table `stepInputs` as input, produce a table `stepOutputs` as output. All tensors inside tables must have the same first dimension `batchSize`. `stepOutputs[1]` should be of shape (`batchSize`, `numTokens`), which denotes the scores in this step. [1]
+  * `nBest` - the `nBest` top hypotheses can be returned after performing beam search. This value must be smaller than or equal to `beamSize`. [1]
+
+]]
+function BeamSearcher:search(beamSize, nBest)
+  assert (nBest <= beamSize)
+  self.nBest = nBest or 1
+  self.beamSize = beamSize or 1
```

Comment:
hmm just for checking `k <= nBest` later one, where `k` is the desired top k-th hypothesis, we must make sure that since the completion criteria for a batch is that the top `nBest` hypotheses all reached EOSs, i.e., if `k > nBest`, the returned hypothesis is invalid since it might have not finished yet.